#313 Specify FT interceptor and the ordering policy (#338)
eclipse/microprofile-fault-tolerance: spec/src/main/asciidoc/fault-tolerance-interceptor.asciidoc

//
// Copyright (c) 2018 Contributors to the Eclipse Foundation
//
// See the NOTICE file(s) distributed with this work for additional
// information regarding copyright ownership.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// You may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// Contributors:
// Emily Jiang

[[fault-tolerance-interceptor]]

== Fault Tolerance Interceptor

> The implementor of MicroProfile Fault Tolerance specification must provide a Fault Tolerance interceptor. The interceptor will be called if any one or many of the above interceptor bindings are specified. The default priority of the interceptor is `Priority.PLATFORM_AFTER+10`, which is `4010`. 

Reviewer: ```suggestion
The implementor of the MicroProfile Fault Tolerance specification must provide a Fault Tolerance interceptor. The interceptor will be called if one or more of the above interceptor bindings are specified. The default priority of this interceptor is `Priority.PLATFORM_AFTER+10`, which is `4010`. 
```

Author: added one sentence.

> The implementor of the MicroProfile Fault Tolerance specification must provide a Fault Tolerance interceptor. This interceptor provides the functionality for each Fault Tolerance annotation. The interceptor will be called if any one or many of the above interceptor bindings are specified. For instance, this interceptor will retry the specified operation if `Retry` annotation was specified on that operation. The default priority of the interceptor is `Priority.PLATFORM_AFTER+10`, which is `4010`. 

Reviewer 2: Another thing that we can change here is this part "The interceptor will be called if any one or many"  would be "The interceptor will be called if one or many of the above interceptor bindings are specified"

> The implementor of the MicroProfile Fault Tolerance specification must provide a Fault Tolerance interceptor. This interceptor provides the functionality for each Fault Tolerance annotation. The interceptor will be called if any one or many of the above interceptor bindings are specified. For instance, this interceptor will retry the specified operation if `Retry` annotation was specified on that operation. The default priority of the interceptor is `Priority.PLATFORM_AFTER+10`, which is `4010`. 

Author: rephrased.

The Fault Tolerance interceptor priority can be configured via MicroProfile Config with the property name of `mp.fault.tolerance.interceptor.priority`. The property value will only be read at application startup. Any subsequent value changes will not take effect until the application restarts.

> If a method is annotated with any of the Fault Tolerance interceptor bindings, it may specify other interceptor bindings. The bounded interceptors will be invoked in the ascending order of the interceptor priority, as specified by https://download.oracle.com/otn-pub/jcp/interceptors-1_2A-mrel3-eval-spec/Intercept.pdf?AuthParam=1541543428_889968d50f535735cbabe343350e09a4[Interceptor Specification^]. 

Reviewer: ```suggestion
A method that is annotated with any of the Fault Tolerance interceptor bindings may also be annotated with other interceptor bindings. The bound interceptors will be invoked in ascending order of interceptor priority, as specified by https://download.oracle.com/otn-pub/jcp/interceptors-1_2A-mrel3-eval-spec/Intercept.pdf?AuthParam=1541543428_889968d50f535735cbabe343350e09a4[Interceptor Specification^]. 
```

Author: fixed

> For an instance, in the following example, MyLogInterceptor will be invoked first, followed by Fault Tolerance interceptor and then MyPrintInterceptor.

Reviewer: ```suggestion
For instance, in the following example, `MyLogInterceptor` will be invoked first, followed by the Fault Tolerance interceptor and then `MyPrintInterceptor`.
```

Author: fixed

> A method, annotated with any of the Fault Tolerance interceptor bindings, may also be annotated with other interceptor bindings. The bound interceptors will be invoked in ascending order of interceptor priority, as specified by https://download.oracle.com/otn-pub/jcp/interceptors-1_2A-mrel3-eval-spec/Intercept.pdf?AuthParam=1541543428_889968d50f535735cbabe343350e09a4[Interceptor Specification^]. If the appliction interceptors are enabled via beans.xml, the interceptors enabled via beans.xml will be invoked after the Fault Tolerance interceptor. For more details, refer to http://docs.jboss.org/cdi/spec/2.0/cdi-spec.html#enabled_interceptors[Interceptor ordering^] in CDI specification.

Reviewer 3: The Oracle links try to track you and will not be accessible.
Please use: https://download.oracle.com/otn-pub/jcp/interceptors-1_2A-mrel3-eval-spec/Intercept.pdf
The JBoss one is ok.

Author: fixed. please double check.
